File: trabalho-tecnologias-remotas-python/graphql_server.py
# graphql_server.py
import strawberry
from typing import List
import data

@strawberry.type
class Song:
    id: int
    title: str
    artist: str

@strawberry.type
class Playlist:
    id: int
    name: str
    user_id: int
    song_ids: List[int]

    @strawberry.field
    def songs(self) -> List[Song]:
        return [Song(**s) for s in data.get_playlist_songs(self.id)]

@strawberry.type
class User:
    id: int
    name: str
    email: str

    @strawberry.field
    def playlists(self) -> List[Playlist]:
        raw_playlists = data.get_playlists(self.id)
        return [Playlist(id=p['id'], name=p['name'], user_id=p['user_id'], song_ids=p['song_ids']) for p in raw_playlists]

@strawberry.type
class Query:
    @strawberry.field
    def users(self) -> List[User]:
        return [User(**u) for u in data.get_users()]

    @strawberry.field
    def songs(self) -> List[Song]:
        return [Song(**s) for s in data.get_songs()]

schema = strawberry.Schema(query=Query)

if __name__ == '__main__':
    import uvicorn
    from strawberry.asgi import GraphQL
    print("GraphQL Server running on port 8002...")
    graphql_app = GraphQL(schema)
    uvicorn.run(graphql_app, host="0.0.0.0", port=8002)

File: trabalho-tecnologias-remotas/graphql-server.ts
// graphql-server.ts
import { ApolloServer } from '@apollo/server';
import { startStandaloneServer } from '@apollo/server/standalone';
import { users, songs, playlists, getPlaylistsByUser, getSongsByPlaylist, User, Playlist } from './data';

// Schema Definition
const typeDefs = `#graphql
  type Song {
    id: Int
    title: String
    artist: String
  }

  type Playlist {
    id: Int
    name: String
    user_id: Int
    song_ids: [Int]
    songs: [Song] # Campo calculado (Resolver)
  }

  type User {
    id: Int
    name: String
    email: String
    playlists: [Playlist] # Campo calculado (Resolver)
  }

  type Query {
    users: [User]
    songs: [Song]
    user(id: Int!): User
  }
`;

// Resolvers
const resolvers = {
    Query: {
        users: () => users,
        songs: () => songs,
        user: (_: any, args: { id: number }) => users.find(u => u.id === args.id),
    },
    User: {
        playlists: (parent: User) => getPlaylistsByUser(parent.id),
    },
    Playlist: {
        songs: (parent: Playlist) => getSongsByPlaylist(parent.id),
    },
};

const startServer = async () => {
    const server = new ApolloServer({
        typeDefs,
        resolvers,
    });

    const { url } = await startStandaloneServer(server, {
        listen: { port: 8002 },
    });

    console.log(`GraphQL Server running on ${url}`);
};

startServer();
